cli/cmd/upgrade/upgrade.go

```diff
@@ -0,0 +1,453 @@
+		return 1
+	}
+
+	// Dry Run should exit here, printing the release's config.
+	if c.flagDryRun {
+		configYaml, err := yaml.Marshal(re.Config)
+		if err != nil {
+			c.UI.Output(err.Error(), terminal.WithErrorStyle())
+			return 1
+		}
+
+		if len(re.Config) == 0 {
+			c.UI.Output("Config: "+string(configYaml), terminal.WithInfoStyle())
+		} else {
+			c.UI.Output("Config:"+"\n"+string(configYaml), terminal.WithInfoStyle())
+		}
+		c.UI.Output("Dry run complete - upgrade can proceed.", terminal.WithSuccessStyle())
+		return 0
+	}
+
+	c.UI.Output("Upgraded Consul into namespace %q", foundNamespace, terminal.WithSuccessStyle())
+
+	return 0
+}
+
+// validateFlags checks that the user's provided flags are valid.
+func (c *Command) validateFlags(args []string) error {
+	if err := c.set.Parse(args); err != nil {
+		return err
+	}
+	if len(c.set.Args()) > 0 {
+		return errors.New("should have no non-flag arguments")
+	}
+	if len(c.flagValueFiles) != 0 && c.flagPreset != defaultPreset {
+		return fmt.Errorf("cannot set both -%s and -%s", flagNameConfigFile, flagNamePreset)
+	}
+	if _, ok := config.Presets[c.flagPreset]; c.flagPreset != defaultPreset && !ok {
+		return fmt.Errorf("'%s' is not a valid preset", c.flagPreset)
+	}
+	if !validLabel(c.flagNamespace) {
+		return fmt.Errorf("'%s' is an invalid namespace. Namespaces follow the RFC 1123 label convention and must "+
+			"consist of a lower case alphanumeric character or '-' and must start/end with an alphanumeric", c.flagNamespace)
+	}
+	duration, err := time.ParseDuration(c.flagTimeout)
+	if err != nil {
+		return fmt.Errorf("unable to parse -%s: %s", flagNameTimeout, err)
+	}
+	c.timeoutDuration = duration
+	if len(c.flagValueFiles) != 0 {
+		for _, filename := range c.flagValueFiles {
+			if _, err := os.Stat(filename); err != nil && os.IsNotExist(err) {
+				return fmt.Errorf("file '%s' does not exist", filename)
+			}
+		}
+	}
+
+	if c.flagDryRun {
+		c.UI.Output("Performing dry run installation.", terminal.WithInfoStyle())
+	}
+	return nil
+}
+
+// validLabel is a helper function that checks if a string follows RFC 1123 labels.
+func validLabel(s string) bool {
+	for i, c := range s {
+		alphanum := ('a' <= c && c <= 'z') || ('0' <= c && c <= '9')
+		// If the character is not the last or first, it can be a dash.
+		if i != 0 && i != (len(s)-1) {
+			alphanum = alphanum || (c == '-')
+		}
+		if !alphanum {
+			return false
+		}
+	}
+	return true
+}
+
+// mergeValuesFlagsWithPrecedence is responsible for merging all the values to determine the values file for the
+// installation based on the following precedence order from lowest to highest:
+// 1. -preset
+// 2. -f values-file
+// 3. -set
+// 4. -set-string
+// 5. -set-file
+// For example, -set-file will override a value provided via -set.
+// Within each of these groups the rightmost flag value has the highest precedence.
+func (c *Command) mergeValuesFlagsWithPrecedence(settings *helmCLI.EnvSettings) (map[string]interface{}, error) {
```

Comment:
This function can be pulled out into common, and referenced by install and upgrade functions since I think its exactly the same in upgrade.go and install.go

Reply:
Great minds think alike. I did this already in the diff branch.